iis-wf/iis-wf-export-actionmanager/src/main/java/eu/dnetlib/iis/wf/export/actionmanager/module/CitationsActionBuilderModuleFactory.java

```diff
@@ -1,118 +1,218 @@
             if (CollectionUtils.isNotEmpty(source.getCitations())) {
                 Result result = new Result();
                 result.setId(source.getDocumentId().toString());
-                
+
                 ExtraInfo extraInfo = new ExtraInfo();
-                extraInfo.setValue(converter.serialize(normalize(source.getCitations())));
+                extraInfo.setValue(citationsExtraInfoSerDe.serialize(
+                        citationEntriesConverter.convert(source.getCitations(), getTrustLevelThreshold())));
                 extraInfo.setName(EXTRA_INFO_NAME);
                 extraInfo.setTypology(EXTRA_INFO_TYPOLOGY);
                 extraInfo.setProvenance(this.getInferenceProvenance());
                 extraInfo.setTrust(StaticConfigurationProvider.ACTION_TRUST_0_9);
                 result.setExtraInfo(Collections.singletonList(extraInfo));
 
                 return result;
-            } else {
-                return null;    
             }
+            return null;
+        }
+
+        public void setCitationsExtraInfoSerDe(CitationsExtraInfoSerDe citationsExtraInfoSerDe) {
+            this.citationsExtraInfoSerDe = citationsExtraInfoSerDe;
+        }
+
+        public void setCitationEntriesConverter(CitationEntriesConverter citationEntriesConverter) {
+            this.citationEntriesConverter = citationEntriesConverter;
+        }
+    }
+
+    /**
+     * Allows citation entry list to be converted to blob citation entry set.
+     */
+    public static class CitationEntriesConverter {
+        private TrustLevelConverter trustLevelConverter = new TrustLevelConverter(InfoSpaceConstants.CONFIDENCE_TO_TRUST_LEVEL_FACTOR);
+        private CitationEntryMatchChecker citationEntryMatchChecker = new CitationEntryMatchChecker();
+        private ConfidenceLevelValidator confidenceLevelValidator = new ConfidenceLevelValidator();
+        private CitationEntryNormalizer citationEntryNormalizer = new CitationEntryNormalizer();
+        private BlobCitationEntryBuilder blobCitationEntryBuilder = new BlobCitationEntryBuilder();
+
+        public SortedSet<BlobCitationEntry> convert(List<CitationEntry> source, Float trustLevelThreshold) {
+            if (source != null) {
+                Float confidenceLevelThreshold = trustLevelConverter.convert(trustLevelThreshold);
+                return source.stream()
+                        .map(citationEntry -> Maps.immutableEntry(citationEntryMatchChecker.check(citationEntry),
+                                citationEntry))
+                        .filter(checkResultAndCitationEntry -> confidenceLevelValidator.validate(
+                                checkResultAndCitationEntry.getKey(), checkResultAndCitationEntry.getValue(), confidenceLevelThreshold))
+                        .map(checkResultAndCitationEntry -> citationEntryNormalizer.normalize(
+                                checkResultAndCitationEntry.getKey(), checkResultAndCitationEntry.getValue()))
```

Comment:
OK, so the steps are the following:
* building the `map<k,v>` where the `k` indicates whether citation has matching result or not and `v` conveys the citation entry
* filtering the entries by `confidenceLevel` and removing all the matches below the threshold (!)
* normalizing the destination document id

Do I get it right?

It's quite nice, decouples the steps having dedicated modules realising each functionality. 
The only problem is it does not work as it should because it removes whole citation entries (including the citation text and external identifiers) for matched citations and this not what we wanted to achieve.

Remember `CitationEntry` conveys not only citation matching outcome but also text and external identifiers so basically what the code does is removing such entry from the list. So as a result you will see missing bibref on the list because the text (and possibly external ids) will not be propagated to the XML produced by the exporter module because WHOLE entry is removed.

What we wanted was as I have described during the call: setting `destinationDocumentId` and `confidenceLevel` to `null` for all the matched cases below the threshold. That's all! No filtering, because we don't want to remove any citation from the citation list.

Good thing: you could simplify the whole logic: no need to filter thus no need to generate `CheckResult` and the whole logic could be encoded in a simple map operation `null`ing two fields (when below threshold) I have mentioned above.

The task is relatively simple, this is why I was surprised it takes so much effort and refactoring. Probably there was still some misunderstanding.

Anyway I am closing this review round and will get back once we get this code on track :)

Reply:
> OK, so the steps are the following:
> 
>     * building the `map<k,v>` where the `k` indicates whether citation has matching result or not and `v` conveys the citation entry
> 
>     * filtering the entries by `confidenceLevel` and removing all the matches below the threshold (!)
> 
>     * normalizing the destination document id
> 
> 
> Do I get it right?

Yes, you're right. 

> It's quite nice, decouples the steps having dedicated modules realising each functionality.
> The only problem is it does not work as it should because it removes whole citation entries (including the citation text and external identifiers) for matched citations and this not what we wanted to achieve.
> 
> Remember `CitationEntry` conveys not only citation matching outcome but also text and external identifiers so basically what the code does is removing such entry from the list. So as a result you will see missing bibref on the list because the text (and possibly external ids) will not be propagated to the XML produced by the exporter module because WHOLE entry is removed.
> 
> What we wanted was as I have described during the call: setting `destinationDocumentId` and `confidenceLevel` to `null` for all the matched cases below the threshold. That's all! No filtering, because we don't want to remove any citation from the citation list.

I've removed filtering so no citation entries are removed and the logic is encoded using only map operations. Destination document id and confidence level are `null`ed for citation entries that contain the outcome of citation matching and confidence level is below threshold.